packages/venia-ui/lib/util/images.js

```diff
@@ -17,23 +17,52 @@ export const imageWidths = {
     XXXLARGE: 2560
 };
 
-const generateURL = (imageURL, mediaBase) => (width, height) => {
-    const url = resourceUrl(imageURL, {
+const generateUrl = (imageURL, mediaBase) => (width, height) =>
+    resourceUrl(imageURL, {
         type: mediaBase,
         width,
         height
     });
 
-    return `${url} ${width}w`;
+export const generateUrlFromContainerWidth = (
+    imageURL,
+    containerWidth,
+    type = 'image-product'
+) => {
+    const intrinsicWidth = window.devicePixelRatio * containerWidth;
+    /**
+     * Using the reduce on imageWidths to find the best width that is
+     * closest to the intrinsicWidth to be used to generate the URL.
+     */
+    const actualWidth = Object.values(imageWidths).reduce((prev, curr) => {
+        if (prev) {
+            return Math.abs(intrinsicWidth - curr) <
+                Math.abs(intrinsicWidth - prev)
+                ? curr
+                : prev;
+        } else {
+            return curr;
+        }
+    }, null);
+
+    return generateUrl(imageURL, type)(
+        actualWidth,
+        actualWidth / DEFAULT_WIDTH_TO_HEIGHT_RATIO
+    );
 };
 
-export const generateSrcset = (imageURL, mediaBase) => {
-    if (!imageURL || !mediaBase) return '';
+export const generateSrcset = (imageURL, type) => {
+    if (!imageURL || !type) return '';
+
+    const generateSrcsetUrl = generateUrl(imageURL, type);
 
-    const generateSrcsetURL = generateURL(imageURL, mediaBase);
     return Object.values(imageWidths)
-        .map(width =>
-            generateSrcsetURL(width, width / DEFAULT_WIDTH_TO_HEIGHT_RATIO)
+        .map(
```

Comment:
I strongly suggest making `imageWidths` a `Map` rather than an object if we're going to iterate over it, especially since we're doing so in multiple places.